Implement AntiCheat System - AntiHacks

kvipka opened the pull request:
**Updated version with log in auth database for fresh TC revision**

-ASH - AntiSpeedHack (+anti teleport hack)
-AFH - AntiFlyHack
-DoubleJump
-Wallclimb
-FakeJumper
-FakeFlyingMode

**Updated by 07 sept 2019:**
- implement ASH control for vehicle and transport/elevators
- implement 2 new movementgenerators : 
* JumpMovementGenerator
* ChargeMovemengGenerator
it will give correct information to server of unit-status and correct unit state. Also it fixed some issues (look info in comments)

This system has been tested on some big projects with big online (project names are hidden). 
**Issues:** No false positives

For some questions : discord - 
https://discord.gg/GcJJfXG

youtube videos : https://www.youtube.com/watch?v=aZP3qm9Ql0Y&list=PLUpIDd4JDWyOOJhJRcfaVL0RaRH1Ra5Xy

Note:
This system using only mathematic operations, without additional manager as AC2, no using additional memory.

Part of code can be shorter, when [PR](https://github.com/TrinityCore/TrinityCore/pull/18771) with ACK-packet handlers will complete.

![ACK 1](https://user-images.githubusercontent.com/2619928/60046307-baecf980-96cf-11e9-9338-84f633d90a1f.png)

Log in auth-database :
![Безымянный](https://user-images.githubusercontent.com/2619928/63114123-b12f9600-bf9c-11e9-8758-8005028c6d0a.png)

kvipka committed:
(Updated TC version)
Implement AntiCheat System - AntiHacks:
-ASH - AntiSpeedHack
-AFH - AntiFlyHack
-DoubleJump
-Wallclimb
-FakeJumper
-FakeFlyingMode

illfated commented:
Ref. #23509 (previous PR with the same content)

Jildor commented:
I'm not sure if log detection in the auth db is a god idea (performance things).
But maybe it's better if you add a check for enable/disable this option by the conf file ;)

kvipka commented:
> performance things

log in database by AccountMgr (another thread) + by encapsulated asynchronous query. I think no reason to worry

jackpoz commented:
how come you create a new PR for the same code ? it makes the review and discussion about the code confusing

illfated commented:
Oh. Is the content here without the changes in the previous PR (which was closed due to file conflicts)?

kvipka commented:
> Oh. Is the content here without the changes in the previous PR (which was closed due to file conflicts)?

"with"

illfated commented:
OK, fair enough. I misinterpreted the comment jackpoz wrote.

jackpoz commented:
> **Issues:** No false positives

does this new PR fix all the false positives reported in the old PR https://github.com/TrinityCore/TrinityCore/pull/23509 ?

kvipka commented:
From previous pr, yes

Infernales commented:
@kvipka I don’t remember if I asked such a question: should anticheat determine the hack of walking on water and the hack of jump height?

kvipka commented:
> @kvipka I don’t remember if I asked such a question: should anticheat determine the hack of walking on water and the hack of jump height?

nope and nope

Infernales commented:
@kvipka Well then, everything is in order.

Infernales commented:
@kvipka False positives:
```
Unit::CheckMovementInfo :  Speed hack detected for Account Id : 1, Player Test
Unit::CheckMovementInfo :  oldX = 16222.885742
Unit::CheckMovementInfo :  oldY = 16260.132812
Unit::CheckMovementInfo :  newX = 16222.971680
Unit::CheckMovementInfo :  newY = 16259.600586
Unit::CheckMovementInfo :  packet distance = 0.539120
Unit::CheckMovementInfo :  available distance = 0.483302
Unit::CheckMovementInfo :  move time = 1249580288.000000
Unit::CheckMovementInfo :  delay sent ptk - recieve pkt (previous) = 0.000000
Unit::CheckMovementInfo :  full delay = 0.000038
Unit::CheckMovementInfo :  diff time = 0.060038
Unit::CheckMovementInfo :  real ping = 0
Unit::CheckMovementInfo :  fake ping = 60

Unit::CheckMovementInfo :  Speed hack detected for Account Id : 1, Player Test
Unit::CheckMovementInfo :  oldX = 16227.772461
Unit::CheckMovementInfo :  oldY = 16255.753906
Unit::CheckMovementInfo :  newX = 16227.333008
Unit::CheckMovementInfo :  newY = 16256.427734
Unit::CheckMovementInfo :  packet distance = 0.804465
Unit::CheckMovementInfo :  available distance = 0.483302
Unit::CheckMovementInfo :  move time = 1249650048.000000
Unit::CheckMovementInfo :  delay sent ptk - recieve pkt (previous) = 0.000000
Unit::CheckMovementInfo :  full delay = 0.000038
Unit::CheckMovementInfo :  diff time = 0.060038
Unit::CheckMovementInfo :  real ping = 0
Unit::CheckMovementInfo :  fake ping = 60
```
Anticheat falsely triggered when the client started on the same machine as the server. Well, a classic of the genre: false positives began after 14 days of the uptime server.

jackpoz commented:
does that happen when the OS updates the clock ?

Infernales commented:
> does that happen when the OS updates the clock ?
No. This happened when starting the torrent client. When I tried to connect to the server from a laptop, the same situation arose there too, but already because of the Windows update.

illfated commented:
Sounds like you created your own local lag situation. Could be useful for testing, provoking lag, but not sure how much it tells us about the PR content.

Infernales commented:
@illfated So the problem is that even after closing the torrent client, it continues to kick from the server.

illfated commented:
I understand that I should avoid guessing what is going on with the code, but could this have anything to do with the size of your logs or the number of logs being written by the servers? (Comparing to how Blizzard servers restart at least once every week, although they usually have a lot of players online.)

Infernales commented:
@illfated It is unlikely, but yes, after restarting the server(TrinityCore), everything works again.

illfated commented:
Fair enough, thank you for confirming that. At least we know it is something that builds up over time (and usually more than a week or 10 days).

kvipka commented:
![WoWScrnShot_090519_172704](https://user-images.githubusercontent.com/2619928/64351136-77353b00-d002-11e9-946c-1c42e8eac3e1.jpg)

Uptime 15 days, no troubles

Infernales commented:
@kvipka Strange, maybe the problem is that I compiled the server with the -DWITH_DYNAMIC_LINKING=1 and -DSCRIPTS="dynamic" parameters? Yes, I forgot to say about it.

kvipka committed:
implement Vehicle speed (fix https://at-wow.org/forum/threads/repo-with-ac.48/#post-92 )

kvipka committed:
implement isMapDisabledForAC for FAKEJUMPER and FAKEFLYINGMODE

Jildor commented:
👍 
Can anyone test vehicle speed with auras tha increase vehicle speed movement like Ulduar - Salvaged Chopper (33062) with 62299 aura?

Aokromes commented on src/server/worldserver/worldserver.conf.dist:
```diff
@@ -4065,3 +4066,116 @@ Metric.OverallStatusInterval = 1
+AntiCheats.forceExcludeMapsid = ""
+
+#
+###################################################################################################
```
missing new line :)

kvipka replied:
> missing new line :)

![tenor](https://user-images.githubusercontent.com/2619928/64429210-297e0880-d0be-11e9-8f79-d1d32c02b66c.gif)

illfated commented on sql/custom/auth/2019_07_24_00_anticheat_logs.sql:
```diff
@@ -0,0 +1,15 @@
+  `position` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
+  `realmId` int(9) NOT NULL,
+  PRIMARY KEY (`id`)
+) ENGINE = InnoDB AUTO_INCREMENT = 0 CHARACTER SET = utf8;
```
Missing NewLine ("No newline at end of file").

kvipka committed:
full implementation ASH for transport and elevators

kvipka committed:
Implement 2 new movementgenerators : JumpMovementGenerator and ChargeMovementGenerator
Fix false/positives for charge/deathgrip and others jump spells (KnockbackFrom and leaps)

kvipka commented:
Implement 2 new movementgenerators : JumpMovementGenerator and ChargeMovementGenerator

Fix false/positives for charge/deathgrip and others jump spells (KnockbackFrom and leaps)

Now Charge and Jump are not virtual movement. And server will know, when movement will end.
Also fixed some issues : 
- troubles with SAI when dk  grip this NPC
- troubles with NPC, when they drops under textures (for example knockback spells like shaman spell)

Infernales commented:
@kvipka Build error:
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Movement/MovementGenerators/FollowMovementGenerator.cpp:204:55: fatal error: 

      no member named 'HasWalkingFlag' in 'Player'

                walk = owner->GetOwner()->ToPlayer()->HasWalkingFlag();

                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^

1 error generated.

src/server/game/CMakeFiles/game.dir/build.make:2527: recipe for target 'src/server/game/CMakeFiles/game.dir/Movement/MovementGenerators/FollowMovementGenerator.cpp.o' failed

make[2]: *** [src/server/game/CMakeFiles/game.dir/Movement/MovementGenerators/FollowMovementGenerator.cpp.o] Error 1
```

kvipka committed:
implement WalkingFlag for Player and Pet Follower

kvipka commented:
Implement WalkingFlag. What is it : 
* When player with Pet are walking - Pet walking too. When Player starting to run - pet starting to run too with owner. In TC it another issue. 

Just too lazy to look for all these issues. =)

kvipka committed:
Main present from AtieshCore to TC :
fixed one the oldest bug with Confuse And Fleeing - when unit are accelerating or teleporting and has asynchronic position for others visitors

kvipka commented:
very lazy to write a description of last commit. Just look this videos : 

Async position Before : 
https://youtu.be/WJe24fYTgkw?list=PLUpIDd4JDWyPuzg5PyVuXoHx6sn7eqgYw
After : 
https://youtu.be/7dmc9-Nsrys?list=PLUpIDd4JDWyPuzg5PyVuXoHx6sn7eqgYw

Fear Before : 
https://youtu.be/vCc8SVWgWsk?list=PLUpIDd4JDWyPuzg5PyVuXoHx6sn7eqgYw
Fear after : 
https://youtu.be/Q9T_tnInBl0?list=PLUpIDd4JDWyPuzg5PyVuXoHx6sn7eqgYw

Just present...

kvipka commented:
some relations : 
https://github.com/TrinityCore/TrinityCore/issues/21255

Jildor commented:
wait, you're fixing a issues related with movement in PR of anticheat implementation? ( 👍 ❤️ )
isn't better split in a separated PR?

illfated commented:
Fair point, because if something in the "fix movement" part needs reverting, it would be much easier to check by removing a specific commit from the main branch.
If you need the fixes to be present for this AntiCheat PR to work, ask for the movement improvement PR to be merged first.

Infernales commented:
@illfated I think he did it because he has to tear out his code in pieces from his fork. In this case, it makes sense to do everything in one PR. Again, sooner or later this PR will be in branch 3.3.5. At least I really hope so.

kvipka commented:
i will not separate charge/jump movegens.
Current TC has virtual realization - and projects has issues with grip/charge - because server don't know real position of player.

Early (and in PR for azerothcore) it handled by HACK : just timer of ACK packets, when ASH are skipping checking in this time.

This hack do possible to use teleporthacks. In this PR i removed this hack, and will not use something like this. 
Only full correct realization.
____

About Confuse/Fleeing improvement - i will not separate it too, it present.
p.s. If you don't want to use it - just don't cherry-pick last commit. As i know - this PR will never merged by TC. And by this reason all ppl can take code, what want from this PR.

illfated commented:
Fair point too. Thank you for clarifying.

kvipka commented:
![Screenshot_20190908-121642](https://user-images.githubusercontent.com/2619928/64486573-5d306e00-d237-11e9-86da-5e2a65599bbb.png)



Gendalf and his magic

Infernales commented:
> Uptime, no trouble with time. Server Ubuntu
Try configuring Cmake with the -DWITH_DYNAMIC_LINKING=1 and -DSCRIPTS="dynamic" options.
Perhaps because of this I had problems with anticheat when I connected to the server from the same PC.

Aokromes commented:
i got reports of gundrak rinos doing "invisible" charges with the latest version of anticheat.

kvipka committed:
Cleared code + fix some false/positives

kvipka commented:
> i got reports of gundrak rinos doing "invisible" charges with the latest version of anticheat.

what the rinos? more info need)

illfated commented:
- https://www.wowhead.com/npc=29931/drakkari-rhino
- https://www.wowhead.com/npc=29838/drakkari-rhino

Aokromes commented:
https://mega.nz/#!x2pBGKLK!5QVu47oYg1jKq_sFTfXMDs6htBw1oNyMDeXp7uKS4kM

kvipka committed:
fixed Charge StopMoving in Gundrak

kvipka committed:
For happy Aokromes

kvipka commented on src/server/worldserver/worldserver.conf.dist:
```diff
@@ -4119,8 +4119,8 @@ AntiCheats.IgnoreControlMovement.Enabled = 1
 
-Appender.Anticheat=2,3,15,Anticheat.log
-Logger.anticheat=3,Console Anticheat
+#Appender.Anticheat=2,3,15,Anticheat.log
```
Now config with additional file with logs commited. All info in auth-database in anticheat_logs.

jackpoz replied:
are old logs ever cleared ?

kvipka replied:
> are old logs ever cleared ?

what you mean? .log files? as you wish. If about data in anticheat_logs - nope.

wowmane commented:
https://github.com/azerothcore/azerothcore-wotlk/pull/2088#issuecomment-531731970

https://github.com/azerothcore/azerothcore-wotlk/pull/2088#issuecomment-529208280

Cobus420 commented:
are there any new failures/bugs here or is this pull ready?

illfated commented:
@Cobus420 : Feedback from individual testers is still useful at this stage, even if this PR is marked as ready ( **[Feedback-PatchFix]** ).

Infernales commented:
@kvipka After 15 days, everything kicks the same way, if you go to the server from the same PC where the server itself is running. 
@anyone Can anyone else confirm this except me?

kvipka commented:
> @kvipka After 15 days, everything kicks the same way, if you go to the server from the same PC where the server itself is running.
> Can anyone else confirm this except me?

can you give answer on this questions :
- do you use windows ?
- if yes, windows has auto correction of time?

Infernales commented:
@kvipka I used three different servers with Arch Linux, Debian and CentOS.

Jildor commented:
When you're play in the same machine that server, you have a very low ms, maybe can be this the cause?

Jildor commented:
oh after 15 days,I did not noticed it, I don't know then xD

Jinnaii commented:
Is this 15 days situation the only thing that stops this PR from merging or or are there a few unresolved false positive left?
In my opinion a Server running for 15+ days is not a realistic example but i may be wrong.

Cobus420 commented:
I got the Server Running for 18 Days with that PR. I dont get any issues with kicks or errors. Running on Windows Server.

Aokromes commented:
it false positives a lot with people with >300ms lag

qyh214 commented:
like @Aokromes  said, lag is 500-700ms,I three jumps(space x 3) then I kick and disconnect with server,

Temporary fix: kick disabled

Jinnaix commented:
Even if this would mean to disable autokick duo to latency false positives, this system is a major improvement to what trinity offers now. Its worth to have for logging and manual banning afterwards.

Is there any way to get and log player latency?

Jildor commented:
yep 
```
GetSession()->GetLatency();
```

kvipka commented:
okay, rewrited algorithm, now should be okay

kvipka committed:
fixed bug with false/positives after more then week uptime

Aokromes commented:
cannot be merged after latest changes.

Conflicting files
src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp

jackpoz commented:
It seems this pr reverts some old MotionMaster.cpp changes too

kvipka committed:
Merge branch '3.3.5' into Antihacks

kvipka commented:
okay, rdy

jackpoz commented:
I would say JumpMovementGenerator and ChargeMovemengGenerator should be in their own PR as they have little to do with anti-cheat

kvipka commented:
Nope, it needed for anticheat

jackpoz commented:
you don't need to remove it, more like open a PR with just those bits so it can be merged before this one

Aokromes commented:
Conflicting files
src/server/game/Spells/Spell.cpp

Aokromes commented:
lol it detected my own player character having 35ms lag.

jackpoz commented:
> 3.500280, distance from packet = 3.50172

maybe it should be less strict than 0.00144 yards

jackpoz commented on src/server/game/Entities/Player/Anticheat.cpp:
```diff
@@ -0,0 +1,396 @@
+            TC_LOG_INFO("anticheat", "oldctime > ptime");
+            delay = 0;
+        }
+        diffPacketdelay = diffPacketdelay * 0.0000000001f;
```
what is the deal with all these 10000000 and 0.0000000001f numbers ? what are they supposed to do ?

kvipka committed:
Merge branch '3.3.5' into Antihacks

kvipka commented:
![Безымянный](https://user-images.githubusercontent.com/2619928/71896395-bc11df80-3164-11ea-8b09-d2b0bc9707c5.png)

11 days of uptime, has no troubles yet

illfated commented:
in case of preparation for merging, I presume there is a missing pch dependency:
```cpp
/home/circleci/project/src/server/game/AI/SmartScripts/SmartScriptMgr.cpp:1508:56:
 fatal error: use of undeclared identifier 'MAX_MOVE_TYPE'
            if (e.action.movementSpeed.movementType >= MAX_MOVE_TYPE)
                                                       ^
1 error generated.

kvipka commented:
> in case of preparation for merging, I presume there is a missing pch dependency:
> 
> ```c++
> /home/circleci/project/src/server/game/AI/SmartScripts/SmartScriptMgr.cpp:1508:56:
>  fatal error: use of undeclared identifier 'MAX_MOVE_TYPE'
>             if (e.action.movementSpeed.movementType >= MAX_MOVE_TYPE)
>                                                        ^
> 1 error generated.
> ```
maybe because this ?
https://github.com/TrinityCore/TrinityCore/commit/be65f7d60e66d4c43f8fe9d779e5b4214afb91bf

smartscripts have not related with anticheats =)

kvipka committed:
Update Anticheat.cpp

jackpoz commented:
>11 days of uptime, has no troubles yet

Did it report any real cheater or any false positive ?

illfated commented:
> > in case of preparation for merging, I presume there is a missing pch dependency:
> > ```c++
> > /home/circleci/project/src/server/game/AI/SmartScripts/SmartScriptMgr.cpp:1508:56:
> >  fatal error: use of undeclared identifier 'MAX_MOVE_TYPE'
> >             if (e.action.movementSpeed.movementType >= MAX_MOVE_TYPE)
> >                                                        ^
> > 1 error generated.
> > ```
> 
> maybe because this ?
> [be65f7d](https://github.com/TrinityCore/TrinityCore/commit/be65f7d60e66d4c43f8fe9d779e5b4214afb91bf)
> 
> smartscripts have not related with anticheats =)

I know, but **ci/circleci: nopch** is currently reacting to the file `src/server/game/AI/SmartScripts/SmartScriptMgr.cpp` inside your PR branch.
I guess it doesn't really matter that much before the final merge takes place.

kvipka commented:
https://www.youtube.com/watch?v=VYy-_pKrTnY

here you can see player with very bad inet connection and big ping interval ("ping horse racing")

Aokromes commented:
Conflicting files
src/server/game/Spells/Spell.cpp

kvipka committed:
Merge branch '3.3.5' into Antihacks

kvipka committed:
upd with the last mmaps changes

Aokromes commented:
btw

git apply ../p3/anticheat
../p3/anticheat:213: trailing whitespace.
// Anticheat System
../p3/anticheat:269: trailing whitespace.
    if (GetAreaId() == 4859) // Area: 4859 (Frozen Throne)
../p3/anticheat:419: trailing whitespace.

../p3/anticheat:1393: trailing whitespace.
        Add(new JumpMovementGenerator<Player>(id, x, y, z, o, speedXY, max_height, hasOrientation));
../p3/anticheat:1970: trailing whitespace.

warning: squelched 7 whitespace errors
warning: 12 lines add whitespace errors.

Aokromes commented:
still to picky
SpeedHack Detected for xxxxxx, normal distance for this time and speed = 3.465000, distance from packet = 3.465672

Infernales commented:
Jumped in the elevator in Mulgore:
```
Player::CheckMovementInfo :  Wall climb hack detected for account Id : 1, Player Test, diffZ = 26.858719, distance = 0.000000, angle = 0.000000, Map = Kalimdor, mapId = 1, X = -5399.479492, Y = -2506.116699, Z = -39.996765
```
When using `.gm fly on` and mount, if you land on the ground, you get this:
```
2020-01-12_16:50:19 Unit::CheckMovementInfo :  Speed hack detected for account Id : 1, Player Test
2020-01-12_16:50:19 Unit::CheckMovementInfo :  oldX = 16228.923828
2020-01-12_16:50:19 Unit::CheckMovementInfo :  oldY = 16254.991211
2020-01-12_16:50:19 Unit::CheckMovementInfo :  newX = 16226.244141
2020-01-12_16:50:19 Unit::CheckMovementInfo :  newY = 16256.898438
2020-01-12_16:50:19 Unit::CheckMovementInfo :  packet distance = 3.289109
2020-01-12_16:50:19 Unit::CheckMovementInfo :  available distance = 1.652000
2020-01-12_16:50:19 Unit::CheckMovementInfo :  oldStime = 401958430
2020-01-12_16:50:19 Unit::CheckMovementInfo :  oldCtime = 402069177
2020-01-12_16:50:19 Unit::CheckMovementInfo :  server time = 401958630
2020-01-12_16:50:19 Unit::CheckMovementInfo :  packet time = 402069412
2020-01-12_16:50:19 Unit::CheckMovementInfo :  diff delay between old ptk and current pkt = 0.001000
2020-01-12_16:50:19 Unit::CheckMovementInfo :  full delay = 0.235000
2020-01-12_16:50:19 Unit::CheckMovementInfo :  diff time = 0.236000
2020-01-12_16:50:19 Unit::CheckMovementInfo :  ping = 9
```

kvipka committed:
Implement NoFallingDmg anti-hack
+ fix false/positive on elevator jumps

kvipka committed:
typo

kvipka committed:
fix .gm fly on + mount = false/positive

kvipka committed:
correct declare

kvipka committed:
safe m_antiNoFallDmg and m_antiNoFallDmgLastChance in SetFallInformation

kvipka committed:
corrected : fixed troubles with ASH and DamageFall on elevators

kvipka committed:
improved previous commit

Aokromes commented:
still:
 git apply ../p3/anticheat
../p3/anticheat:275: trailing whitespace.
    if (GetAreaId() == 4859) // Area: 4859 (Frozen Throne)
../p3/anticheat:1660: trailing whitespace.
        Add(new JumpMovementGenerator<Player>(id, x, y, z, o, speedXY, max_height, hasOrientation));
../p3/anticheat:2246: trailing whitespace.

../p3/anticheat:2260: trailing whitespace.

../p3/anticheat:2345: trailing whitespace.

warning: squelched 4 whitespace errors
warning: 9 lines add whitespace errors.

Aokromes commented:
lately i am getting reports of people taking damage on land from flying mounts.

Aokromes commented:
https://gist.github.com/Aokromes/3488abafe772e5b10a480cd9ba54c209

plrMover->SetFallInformation(plrMover->GetPositionZ());

is crashing core acording shauren.

kvipka commented:
good catch!

kvipka committed:
correction from Aokromes info

kvipka committed:
noone have seen it

Aokromes commented:
i am still getting damage reports on landing.

kvipka committed:
reworked old SetFallInformation in new ResetFallingData and return back UpdateFallInformationIfNeed

kvipka commented:
> i am still getting damage reports on landing.

pls more info, how to reproduce in future,maybe screens

Aokromes commented:
https://mega.nz/#!xmZjiSYA!u0t8RjWyQN6lw2N1OUzWpI8-bGOE2ImVEX9jwoGYsLE

kvipka commented:
Good, what addon of frame if nit a secret? :)

Aokromes commented:
this false positives a lot.

MovementHandler::NoFallingDamage by Account id : , Player 
MovementHandler::NoFallingDamage by Account id : , Player 
MovementHandler::NoFallingDamage by Account id : , Player

kvipka commented:
It interesting. This announce will working only when player, which have flag of falling_far, have lost this flag without sending land or swim opcode and without calling function ResetFallingInfo.In this time server start timer, while will waiting this opcode from player. When timer expired, server will send this announce and kick player. How player can receive this announce without hack idk. Can you ask him to do screenshots or better to do video?

Aokromes commented:
i asked the guy bug he doesnt' remember exact place (i asked him 3h after got reported)

kvipka committed:
Merge branch '3.3.5' into Antihacks

kvipka committed:
added sync position at start jump

Aokromes commented:
i still get a lot of false positives of 
MovementHandler::NoFallingDamage by Account id : , Player
and of players with high latency.

kvipka committed:
need more timer for laggers with ping > 1000ms-2000ms

Aokromes commented:
meh https://i.imgur.com/azHF9n0.jpg

kvipka committed:
fix false/positive NoFallingDmg when druid instant fly-form while falling

kvipka committed:
a little improvement

kvipka committed:
fix rarely false/positives on edge texture

kvipka committed:
fix DeathGrip/broken on NPC-casters

kvipka commented:
fixed issue when dk trying to do deathgrip on npc, but unit starting cast (for example scarlet priest).

kvipka committed:
for happy Aokromes temporary hack

jackpoz commented:
This PR has a lot of changes all around the core and it might never be actually merged. What will most likely happen is:
1) The PR is left open here, waiting and waiting
2) The PR is pushed at https://github.com/TrinityCore/TrinityCoreCustomChanges as a branch that anyone can choose to use , listed at https://github.com/TrinityCore/TrinityCoreCustomChanges/wiki

Do you have any preference out of these 2 choices @kvipka ?

kvipka commented:
@jackpoz i tihnk 1 variant are better, it's not a custom feature, just security.

Aokromes commented:
i think this breaks flame leviathan fight, it doesn't even agro after breaking door.

kvipka commented:
> i think this breaks flame leviathan fight, it doesn't even agro after breaking door.

how movement check can be related with bug with aggro-system ?

jackpoz commented:
Like https://github.com/TrinityCore/TrinityCore/commit/3d11de239c294259bc9b6ec5c2638050fea0bf65 maybe ?

Aokromes commented:
> > i think this breaks flame leviathan fight, it doesn't even agro after breaking door.
> 
> how movement check can be related with bug with aggro-system ?

well, remind invisible charges on gundrak xd this is the only big patch i have on my server non-present on tc test server.

Aokromes commented:
> > i think this breaks flame leviathan fight, it doesn't even agro after breaking door.
> 
> how movement check can be related with bug with aggro-system ?

the only custom patch on this test server.

https://mega.nz/#!Mm5VyYYK!dvCNHg0W14pGo336OnhXRXtfl2xdSJfGZkhEwmSa2d4

Aokromes commented:
[11:35:24] AntiCheat: Wallclimb Detected for Account id : , Player , diffZ = 22.276047, distance = 0.000000, angle = 0.000000, Map = Gundrak, mapId = 604, X = 1834.008911, Y = 858.692688, Z = 127.811676
i was told he dispelled priest levitate while falling.

Rushor commented:
@kvipka do you have anything against PQR ?

Aokromes commented:
a warrior charging while being stunned on the road on gundrak:
[21:46:01] AntiCheat: NoFallingDamage Hack Detected for Account : , Player :

Aokromes commented:
it looks it also breaks the light of dawn quest.

jackpoz commented:
This PR is more of its own separate project that will require fixes and tuning over the years, just like TC. 
We cannot expect it to be perfect but at the same time this means we cannot merge it.
It can stay here but I really think https://github.com/TrinityCore/TrinityCoreCustomChanges would be a better place (and I might ask to move it there sometime in the future)

Lillecarl commented:
@kvipka how about making a pullrequest against trinitycorecustomchanges instead?

Aokromes commented:
> Iirc there already is a branch for anti-cheat.

that's the anticheat i used to maintain, very diferent from this one.

MaxtorCoder commented:
Ooohh, my bad then.

warcrafte commented:
Bug Spell : [Master's Call] = If use, Pet Goes and you need again to use call pet spell.

Class Hunter

peycho commented:
I would love to see this project revived and added to TC. Yes, I know is not perfect (yet), but it might be.

illfated commented:
> I would love to see this project revived and added to TC. Yes, I know is not perfect (yet), but it might be.

Sure, but how would you feel if you find there are _lots_ of false positives and other bugs that makes it close to unusable for you, then finding out it takes several months to fix those bugs?

Jinnaix commented:
To be honest, whoever runned a public server once knows that cheaters/hackers are a problem.
TrinityCore provides only basic warden defense that helps against stoneage stuff.
This system here might have flaws, maybe alot, but is a really good counter against the most popular cheats and hacks. 
You shouldn't use it on autokick/ban, but instead as a logger where you manually take actions afterwards. A flyhacker / teleporter / speed hacker doesn't only use his stuff for 2 minutes a day so you will be able to identify them.

Aokromes commented:
> This system here might have flaws, maybe alot, but is a really good counter against the most popular cheats and hacks.
> You shouldn't use it on autokick/ban, but instead as a logger where you manually take actions afterwards. A flyhacker / teleporter / speed hacker doesn't only use his stuff for 2 minutes a day so you will be able to identify them.

The problem is when it breaks playability, ie flame leviathan stops working and so.

Jinnaix commented:
Yes, these problems needs to be fixed of course

kvipka committed:
Merge branch '3.3.5' into Antihacks

jackpoz commented on src/server/game/Entities/Player/Player.h:
```diff
@@ -1695,6 +1695,47 @@ class TC_GAME_API Player : public Unit, public GridObject<Player>
+        /*********************************************************/
+        /***                  ANTICHEAT SYSTEM                   ***/
+        /*********************************************************///
+        // AntiCheat
```
It would be better to move all the anticheat stuff to its own class and call it from Player

lastdarken commented:
Hi

error build

src/server/game/Accounts/AccountMgr.cpp:508:22: fatal error: cannot form pointer to deduced class template specialization type
    PreparedStatement* stmt = LoginDatabase.GetPreparedStatement(LOGIN_INS_ACCOUNT_ANTICHEAT);
                     ^

illfated commented:
@lastdarken : if this is something you only get from adding this PR content to your source, you should address your issue to the PR owner.

Infernales commented:
@illfated I talked with the author of this PR. He has no motivation for fixing conflicts and compilation errors after new commits in TC, since this PR will never be merged with the 3.3.5 branch. This PR will only be updated if the author adds any new features. Thus, we take shovels and remove all the resulting shit on our own.
@lastdarken 
```
diff --git a/src/server/game/Accounts/AccountMgr.cpp b/src/server/game/Accounts/AccountMgr.cpp
index f9422a2..f60a6d5 100644
--- a/src/server/game/Accounts/AccountMgr.cpp
+++ b/src/server/game/Accounts/AccountMgr.cpp
@@ -505,7 +505,7 @@
 
 void AccountMgr::RecordAntiCheatLog(uint32 accountId, std::string const& name, std::string const& description, std::string const& position, uint32 realmid)
 {
-    PreparedStatement* stmt = LoginDatabase.GetPreparedStatement(LOGIN_INS_ACCOUNT_ANTICHEAT);
+    LoginDatabasePreparedStatement* stmt = LoginDatabase.GetPreparedStatement(LOGIN_INS_ACCOUNT_ANTICHEAT);
     stmt->setUInt32(0, accountId);
     stmt->setString(1, name);
     stmt->setString(2, description);
diff --git a/src/server/game/Movement/MovementGenerators/ChargeMovementGenerator.cpp b/src/server/game/Movement/MovementGenerators/ChargeMovementGenerator.cpp
index 7bd7a5b..94aa083 100644
--- a/src/server/game/Movement/MovementGenerators/ChargeMovementGenerator.cpp
+++ b/src/server/game/Movement/MovementGenerators/ChargeMovementGenerator.cpp
@@ -79,7 +79,7 @@
     owner->SetIsCharging(true);
     // Call for creature group update
     if (Creature* creature = owner->ToCreature())
-        creature->SignalFormationMovement(Position(_x, _y, _z), _movementId);
+        creature->SignalFormationMovement();
 }
 
 template<class T>
@@ -184,7 +184,7 @@
     owner->SetIsCharging(true);
     // Call for creature group update
     if (Creature* creature = owner->ToCreature())
-        creature->SignalFormationMovement(Position(_x, _y, _z), _movementId);
+        creature->SignalFormationMovement();
 }
 
 template<class T>
```
I did not deal with the conflict after the cfd63c3 commit because I don't have time for this.

illfated commented:
OK, fair point. I see what you mean. Never mind,

kvipka committed:
Fixed wrong calculation by hack in GetClosePoint and implement new GetTheClosestPoint
Fixed false/positives bu fake_jumper

kvipka committed:
Merge branch '3.3.5' into Antihacks

kvipka committed:
some clear

kvipka committed:
for circleci

kvipka committed:
a little correction

kvipka committed:
upd with KickPlayer info

kvipka committed:
LoginDatabasePreparedStatement insted of PreparedStatement

kvipka committed:
tabs

kvipka committed:
tabs again

kvipka committed:
_owner->GetGUID().GetCounter() rewrited in _owner->GetGUID().ToString().c_str()

kvipka committed:
typo

kvipka committed:
typo2

kvipka committed:
typo3

peycho commented:
@kvipka Thanks for the updates.

lastdarken commented:
@kvipka  Thanks.

jackpoz commented:
Added a link to https://github.com/kvipka/TrinityCore/tree/Antihacks at https://github.com/TrinityCore/TrinityCoreCustomChanges/wiki so users can find this branch.

Due to the reasons explained before, this PR cannot be merged to TC. Users are free to merge it locally and report issues to the author.

Thanks for the effort spent into it.

Shadowfiende commented:
@kvipka

Please update to last version AC !


https://github.com/azerothcore/azerothcore-wotlk/pull/2088

Shadowfiende commented:
Pro coder @kvipka  please updated for azerothcore
